### PR TITLE
xsdk: Very preliminary xsdk.py definition for installing xSDK packages

### PR DESCRIPTION
**Pull request by BarrySmith**
This is a replacement for pull request https://github.com/LLNL/spack/pull/2035

After rebase over latest develop branch

I had to make a new pull request because I did not change the branch name in my fork before adding code.

### CODE REVIEW

**Comment by tgamblin** [comment]
@BarrySmith: even if you rebase in your fork's `develop` branch, you can still force-push and it'll show up in the PR.  It's pretty nice because the old commits will go away and the new ones will appear lower in the conversation thread.

**Comment by BarrySmith** [comment]
@tgamblin Agreed. The problem is I made a mistake of pushing when my fork was in a state with the develop branch in my fork matching the develop branch in the LLNL repository (I had move my changes to another branch). As soon as I pushed github noted the pull request branch matched the branch in the LLNL repository and so marked the pull request as complete.  I couldn't find a way to reopen a complete pull request (which makes sense). So my intention was not to start a new pull request, I just did something that prematurely closed my original pull request.

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 1]
```diff
@@ -85,8 +88,12 @@ class Petsc(Package):
     # conflict in headers see
     # https://bitbucket.org/petsc/petsc/src/90564b43f6b05485163c147b464b5d6d28cde3ef/config/BuildSystem/config/packages/hypre.py
     depends_on('hypre~internal-superlu', when='+hypre+mpi~complex')
-    depends_on('superlu-dist@:4.3', when='@:3.6.4+superlu-dist+mpi')
-    depends_on('superlu-dist@5.0.0:', when='@3.7:+superlu-dist+mpi')
+    depends_on('hypre@develop~internal-superlu', when='@develop+hypre+mpi~complex')
+    depends_on('superlu-dist@:4.3', when='@3.6.4+superlu-dist+mpi')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: i don't think you want to change `:3.6.4` to `3.6.4` here.

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 2]
```diff
@@ -85,8 +88,12 @@ class Petsc(Package):
...
+    depends_on('superlu-dist@5.0.0', when='@3.7+superlu-dist+mpi')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: same here, `depends_on('superlu-dist@5.0.0:', when='@3.7:+superlu-dist+mpi')` looked correct to me.

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 3]
```diff
@@ -85,8 +88,12 @@ class Petsc(Package):
     # conflict in headers see
     # https://bitbucket.org/petsc/petsc/src/90564b43f6b05485163c147b464b5d6d28cde3ef/config/BuildSystem/config/packages/hypre.py
     depends_on('hypre~internal-superlu', when='+hypre+mpi~complex')
-    depends_on('superlu-dist@:4.3', when='@:3.6.4+superlu-dist+mpi')
-    depends_on('superlu-dist@5.0.0:', when='@3.7:+superlu-dist+mpi')
+    depends_on('hypre@develop~internal-superlu', when='@develop+hypre+mpi~complex')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: i think we generally do not enforce `develop` variants for packages which are build with `develop`. So i would remove this completely. If you need to build `petsc` with `hypre@develop` it should be fine to do `spack install petsc ^hypre@develop`.

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 4]
```diff
@@ -85,8 +88,12 @@ class Petsc(Package):
...
+    depends_on('superlu-dist@:4.3', when='@3.6.4+superlu-dist+mpi')
+    depends_on('superlu-dist@5.0.0', when='@3.7+superlu-dist+mpi')
+    depends_on('superlu-dist@develop', when='@develop+superlu-dist+mpi')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: same here.

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 5]
```diff
@@ -85,8 +88,12 @@ class Petsc(Package):
...
+    depends_on('trilinos@12.6.2', when='@3.7+trilinos+mpi')
+    depends_on('trilinos@develop', when='@develop+trilinos+mpi')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: and here.

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 6]
```diff
@@ -85,8 +88,12 @@ class Petsc(Package):
     # conflict in headers see
     # https://bitbucket.org/petsc/petsc/src/90564b43f6b05485163c147b464b5d6d28cde3ef/config/BuildSystem/config/packages/hypre.py
     depends_on('hypre~internal-superlu', when='+hypre+mpi~complex')
-    depends_on('superlu-dist@:4.3', when='@:3.6.4+superlu-dist+mpi')
-    depends_on('superlu-dist@5.0.0:', when='@3.7:+superlu-dist+mpi')
+    depends_on('hypre@develop~internal-superlu', when='@develop+hypre+mpi~complex')
+    depends_on('superlu-dist@:4.3', when='@3.6.4+superlu-dist+mpi')
+    depends_on('superlu-dist@5.0.0', when='@3.7+superlu-dist+mpi')
+    depends_on('superlu-dist@develop', when='@develop+superlu-dist+mpi')
+    depends_on('trilinos@12.6.2', when='@3.7+trilinos+mpi')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: does `petsc` use `Trilinos` anyhow? Why exactly `12.6.2`? It should probably also be `@3.7:`.

File: var/spack/repos/builtin/packages/trilinos/package.py [hunk 7]
```diff
@@ -107,6 +110,7 @@ def url_for_version(self, version):
     depends_on('scalapack', when='+mumps')
     depends_on('superlu-dist@:4.3', when='@:12.6.1+superlu-dist')
     depends_on('superlu-dist', when='@12.6.2:+superlu-dist')
+    depends_on('superlu-dist@develop', when='@develop:+superlu-dist')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/trilinos/package.py]: same here.

**Comment by davydden** [comment]
just to clarify on those `develop`: the fact that a user installs `petsc@develop` (i.e. from git) does not mean that he HAS TO use `hypre@develop`, but of course he CAN use it as long as `@develop` version of hypre satisfy `depends_on` in `petsc` package. So that's why i think all those extra lines on `develop` should go.

File: var/spack/repos/builtin/packages/petsc/package.py (hunk 2 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: i.e. you probably don't want to modify this line at all.

File: var/spack/repos/builtin/packages/petsc/package.py (hunk 6 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: also it's not used in config below. Installation page of `petsc` https://www.mcs.anl.gov/petsc/documentation/installation.html does not mention `trilinos`.

**Comment by davydden** [comment]
btw, if the point is to control `develop` vs `release` in `Xsdk` package, then it will already work just fine based on `depends_on` inside `Xsdk` without those extra `depends_on(hypre@develop)` you are suggesting for `petsc`. That's because Spack will create a single graph of packages which satisfies all requirements starting from the root node (presumably `Xsdk`). So if you say there you need `develop` for all linear algebra, they will be used within the whole graph.

**Comment by tgamblin** [comment]
@BarrySmith: I actually agree with @davydden here -- propagating the `@develop` version constraint to `hypre` within `petsc` limits the user in that they can't specify an arbitrary combination of versions on the command line anymore.  Doing this would cause an error:

`spack install petsc@develop ^hypre@2.10.1`

I think the above might be useful for testing (e.g. if you know there is some bug/regression with a particular version of something).

But, if `xsdk@develop` already depends on `petsc@develop` and `hypre@develop`, you can be sure that there's an easy way to get the latest dev stack.

One thing I've thought about is making the package constraints soft, so a user could override them if they wanted to (maybe `petsc@develop` just "prefers" `hypre@develop`).  The concretization solver isn't quite smart enough to do that at the moment though.

**Comment by davydden** [comment]
> there's an easy way to get the latest dev stack.

one can also add those constraints in `~/.spack/packages.yaml`, which I use personally use quite a lot.

File: var/spack/repos/builtin/packages/superlu-dist/package.py [hunk 8]
```diff
@@ -32,6 +33,8 @@ class SuperluDist(Package):
     homepage = "http://crd-legacy.lbl.gov/~xiaoye/SuperLU/"
     url = "http://crd-legacy.lbl.gov/~xiaoye/SuperLU/superlu_dist_4.1.tar.gz"
 
+    version('develop', git='https://github.com/xiaoyeli/superlu_dist', tag='master')
+    version('xsdk-0.1.0', git='https://github.com/xiaoyeli/superlu_dist', commit='0b5369f')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/superlu-dist/package.py]: FYI,  you can't build `petsc@3.7:` with `superlu-dist@xsdk-0.1.0` because `petsc` `depends_on('superlu-dist@5.0.0:', when='@3.7:+superlu-dist+mpi')` whereas `xsdk-0.1.0 < 5.0.0`. 

Workaround is https://github.com/LLNL/spack/pull/1983 + `develop.xsdk.0.1.0` or alike.

File: lib/spack/env/cc [hunk 9]
```diff
@@ -320,7 +320,7 @@ esac
 #
 # Unset pesky environment variables that could affect build sanity.
 #
-unset LD_LIBRARY_PATH
+#unset LD_LIBRARY_PATH
```
Comment (davydden) [inline lib/spack/env/cc]: i think this does not belong to this PR and requires a separate discussion. @alalazo @tgamblin @adamjstewart ping.

I suggest creating a separate PR with this change only, or at least remove it from this PR.

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 10]
```diff
@@ -86,7 +89,8 @@ class Petsc(Package):
     # https://bitbucket.org/petsc/petsc/src/90564b43f6b05485163c147b464b5d6d28cde3ef/config/BuildSystem/config/packages/hypre.py
     depends_on('hypre~internal-superlu', when='+hypre+mpi~complex')
     depends_on('superlu-dist@:4.3', when='@:3.6.4+superlu-dist+mpi')
-    depends_on('superlu-dist@5.0.0:', when='@3.7:+superlu-dist+mpi')
+    depends_on('superlu-dist@5.0.0:', when='@3.7.0:+superlu-dist+mpi')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: there is no need to write `@3.7.0:` the original code `@3.7:` will work just fine.

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 11]
```diff
@@ -86,7 +89,8 @@ class Petsc(Package):
...
+    depends_on('trilinos@12.6.2:', when='@3.7.0:+trilinos+mpi')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: out of curiosity, how does `petsc@3.7.0:` use `trilinos`?

File: lib/spack/env/cc (hunk 9 above)
Comment (BarrySmith) [inline lib/spack/env/cc]: Sorry this was not a "real change" for the pull request, just some intermediate mucking around.
